src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+
+grpc_core::TraceFlag grpc_trace_ares_resolver(false,
+                                              "event_engine_ares_resolver");
+
+namespace {
+
+absl::Status AresStatusToAbslStatus(int status, absl::string_view error_msg) {
+  switch (status) {
+    case ARES_ECANCELLED:
+      return absl::CancelledError(error_msg);
+    case ARES_ENOTIMP:
+      return absl::UnimplementedError(error_msg);
+    case ARES_ENOTFOUND:
+      return absl::NotFoundError(error_msg);
+    default:
+      return absl::UnknownError(error_msg);
+  }
+}
+
+EventEngine::Duration calculate_next_ares_backup_poll_alarm_duration() {
+  // An alternative here could be to use ares_timeout to try to be more
+  // accurate, but that would require using "struct timeval"'s, which just
+  // makes things a bit more complicated. So just poll every second, as
+  // suggested by the c-ares code comments.
+  return std::chrono::seconds(1);
+}
+
+bool IsIpv6LoopbackAvailable() {
+#ifdef GRPC_POSIX_SOCKET_ARES_EV_DRIVER
+  return PosixSocketWrapper::IsIpv6LoopbackAvailable();
+#elif defined(GRPC_WINDOWS_SOCKET_ARES_EV_DRIVER)
+  // TODO(yijiem): make this portable for Windows
+  return true;
+#else
+#error "Unsupported platform"
+#endif
+}
+
+absl::Status SetRequestDNSServer(absl::string_view dns_server,
+                                 ares_channel channel) {
+  GRPC_ARES_RESOLVER_TRACE_LOG("Using DNS server %s", dns_server.data());
+  grpc_resolved_address addr;
+  struct ares_addr_port_node dns_server_addr = {};
+  if (grpc_parse_ipv4_hostport(dns_server, &addr, /*log_errors=*/false)) {
+    dns_server_addr.family = AF_INET;
+    struct sockaddr_in* in = reinterpret_cast<struct sockaddr_in*>(addr.addr);
+    memcpy(&dns_server_addr.addr.addr4, &in->sin_addr, sizeof(struct in_addr));
+    dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+    dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+  } else if (grpc_parse_ipv6_hostport(dns_server, &addr,
+                                      /*log_errors=*/false)) {
+    dns_server_addr.family = AF_INET6;
+    struct sockaddr_in6* in6 =
+        reinterpret_cast<struct sockaddr_in6*>(addr.addr);
+    memcpy(&dns_server_addr.addr.addr6, &in6->sin6_addr,
+           sizeof(struct in6_addr));
+    dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+    dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+  } else {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Cannot parse authority: ", dns_server));
+  }
+  int status = ares_set_servers_ports(channel, &dns_server_addr);
+  if (status != ARES_SUCCESS) {
+    return AresStatusToAbslStatus(
+        status, absl::StrCat("c-ares status is not ARES_SUCCESS: ",
+                             ares_strerror(status)));
+  }
+  return absl::OkStatus();
+}
+
+struct QueryArg {
+  QueryArg(AresResolver* ar, int id, absl::string_view name)
+      : ares_resolver(ar), callback_map_id(id), qname(name) {}
+  AresResolver* ares_resolver;
```

Comment:
Does this need to hold a `RefCountedPtr<>` instead of a raw pointer, so that we're holding a ref for every outstanding query?  Or is that somehow handled implicitly by the polling code?